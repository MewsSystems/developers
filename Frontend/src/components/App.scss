html,
body {
  margin: 0;
  padding: 0;
}

html {
  background: #000;
  color: #fff;
}

// react-paginate library doesn't provide api to replace its componets
// so I didn't find a way to use styled components with it
.Search-Page-Pagination {
  display: flex;
  list-style: none;
  justify-content: center;
  margin: 2rem 0;

  .previous,
  .next {
    display: none;
  }

  li {
    border: 1px solid #fff;
    margin: 0 0.25rem;
    padding: 0;
    cursor: pointer;

    &.break {
      border: none;
    }

    &.selected {
      background: #fff;
      color: #000;
      border-color: #000;
    }
  }

  a {
    display: block;
    padding: 0.5rem;
  }
}
