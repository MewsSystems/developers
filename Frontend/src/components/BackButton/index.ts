export { default } from './BackButton';
export * from './styled';
