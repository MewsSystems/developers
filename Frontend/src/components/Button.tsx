import React from 'react'
import styled, { css } from 'styled-components'
import { BORDER_RADIUS, BOX_SHADOW } from 'constants/index'
import { COLORS } from 'constants/colors'
import { darken } from 'polished'

export const StyledButton = styled.button<
  Pick<ButtonProps, 'disabled' | 'variant' | 'fullWidth'>
>`
  width: ${({ fullWidth }) => fullWidth && '100%'};
  min-width: 12rem;
  padding: 1rem;
  text-align: center;
  text-transform: uppercase;
  border: 0;
  border-radius: ${BORDER_RADIUS.MEDIUM};
  box-shadow: ${BOX_SHADOW.MEDIUM};

  ${({ disabled, variant }) =>
    disabled
      ? css`
          background: ${COLORS.GRAY};
          color: ${COLORS.DARK_GRAY};
        `
      : css`
          background: ${variant === 'primary' ? COLORS.BLUE : COLORS.RED};
          color: ${COLORS.WHITE};
          cursor: pointer;
          transition: background ease-in 0.5s;

          :hover {
            background: ${variant === 'primary'
              ? darken(0.1, COLORS.BLUE)
              : darken(0.1, COLORS.RED)};
            transition: background ease-in 0.5s;
          }
        `}
`

export interface ButtonProps {
  variant?: 'primary' | 'secondary'
  disabled?: boolean
  fullWidth?: boolean
  onClick?: (event: React.MouseEvent<HTMLButtonElement>) => void
  children?: React.ReactNode | React.ReactNodeArray
}

export const Button: React.FC<ButtonProps> = ({
  variant = 'primary',
  disabled,
  fullWidth,
  onClick,
  children,
}) => (
  <StyledButton
    variant={variant}
    disabled={disabled}
    fullWidth={fullWidth}
    onClick={onClick}
  >
    {children}
  </StyledButton>
)
