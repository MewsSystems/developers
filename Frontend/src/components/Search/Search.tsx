import React, { useEffect } from 'react';
import { useDispatch, useSelector, shallowEqual } from 'react-redux';
import ReactPaginate from 'react-paginate';

import scrollToTop from '../../utils/scrollToTop';
import { RootReducer } from '../../reducers';
import { fetchTopRatedMovies, searchMovies, handleSearchChange } from '../../actions/search';
import config from '../../config';

import { Title, SearchResultsWrapper, SearchInput, MovieLink, MoviePosterImage } from './Styled';

const Search = () => {
  const dispatch = useDispatch();
  const loading = useSelector((state: RootReducer) => state.search.loading);
  const searchValue = useSelector((state: RootReducer) => state.search.value);
  const list = useSelector((state: RootReducer) => state.search.list);
  const pagination = useSelector((state: RootReducer) => state.search.pagination, shallowEqual);

  useEffect(() => {
    // there is a warning is browser console about this, but this is OK
    // I don't want to rerender on every loading change
    // it just controls first fetch and does not fetch when we come back from movie page
    if (loading) {
      dispatch(fetchTopRatedMovies());
    }
  }, [dispatch]);

  const onSearchChange = (event: React.ChangeEvent) => {
    event.preventDefault();
    const value = (event.target as HTMLInputElement).value;
    dispatch(handleSearchChange(value));
    if (value) {
      dispatch(searchMovies(value));
    } else {
      dispatch(fetchTopRatedMovies());
    }
  };

  const handlePageChange = ({ selected }) => {
    if (searchValue) {
      dispatch(searchMovies(searchValue, selected + 1));
    } else {
      dispatch(fetchTopRatedMovies(selected + 1));
    }
    scrollToTop();
  };

  return (
    <div>
      <SearchInput onChange={onSearchChange} placeholder="Type to search..." type="search" value={searchValue} />
      {!searchValue && <Title>Top Rated Movies</Title>}
      <SearchResultsWrapper>
        {list.map((it) => (
          <MovieLink key={it.id} to={`/movie/${it.id}`}>
            {it.poster_path ? (
              <MoviePosterImage src={`${config.IMAGE_BASE_URL}${it.poster_path}`} alt={it.title} title={it.title} />
            ) : (
              <div>{it.title}</div>
            )}
          </MovieLink>
        ))}
      </SearchResultsWrapper>

      <ReactPaginate
        containerClassName="Search-Page-Pagination"
        disableInitialCallback
        forcePage={pagination.page - 1}
        initialPage={0}
        marginPagesDisplayed={2}
        onPageChange={handlePageChange}
        pageCount={pagination.total}
        pageRangeDisplayed={5}
      />
    </div>
  );
};

export default Search;
