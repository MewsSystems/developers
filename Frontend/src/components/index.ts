import styled from "styled-components";
import { Link } from "react-router-dom";

export const Layout = styled.div`
  display: grid;
  grid-template-columns: 350px 1fr 350px;
  height: 100vh;
`;

export const CenterColumn = styled.div`
  grid-column: 2;
  padding: 20px;
`;

export const Card = styled.div`
  box-shadow: 0 8px 16px 0 rgba(0,0,0,0.2);
  display: flex;
  padding: 10px;
  color: gray;
  font: 16px/1.5 Helvetica, Verdana, sans-serif;
`

export const Content = styled.div`
  padding: 2px 16px;
`

export const Title = styled.div`
  display: block;
  margin: 10px 0;
  font-weight: bold;
  color: black;
  font-size:20px;
`;

export const SubTitle = styled.div`
  font-size:14px;
  color:grey;
  margin:2px 0;
`;

export const Paragraph = styled.p`
  line-height: 18px;
  margin: 5px;
`;

export const ErrorMessage = styled.div`
  border:1px solid red;
  padding:10px;
  color:red;
  font: 18px Helvetica, Verdana, sans-serif;
`



export const StyledInput = styled.input`
  background-color: transparent;
  border: none;
  border-bottom: 1px solid #9e9e9e;
  border-radius: 0;
  outline: none;
  height: 3rem;
  width: 100%;
  font-size: 16px;
  margin: 0 0 8px 0;
  padding: 0;
  box-shadow: none;
  box-sizing: content-box;
  transition: box-shadow .3s, border .3s;
`;

export const StyledLabel = styled.label`
  width:100%;
  color: #9e9e9e;
  position: absolute;
  top: -15px;
  left: 0;
  font-size: 1rem;
  cursor: text;
  transition: transform .2s ease-out, color .2s ease-out;
  text-align: initial;
  &::after {
    display: block;
    content: "";
    position: absolute;
    top: 100%;
    left: 0;
    opacity: 0;
    transition: .2s opacity ease-out, .2s color ease-out;
  }
`

export const Button = styled.button`
  text-decoration: none;
  color: #fff;
  background-color: #26a69a;
  text-align: center;
  letter-spacing: .5px;
  transition: background-color .2s ease-out;
  cursor: pointer;
  font-size: 14px;
  outline: 0;
  border: none;
  border-radius: 2px;
  display: inline-block;
  height: 36px;
  line-height: 36px;
  padding: 0 16px;
  text-transform: uppercase;
  vertical-align: middle;
  &:disabled {
    background-color:#ccc;
  }
`;

export const StyledLink = styled(Link)`
  text-decoration: none;
  color: #26a69a;
`;
