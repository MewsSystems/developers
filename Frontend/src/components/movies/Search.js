import React from 'react';

const Search = ({ change }) => {

  return (
    <div>
      <input
        id="search"
        type="search"
        placeholder="Search movies..."
        onChange={e => change(e.target.value)}/>
    </div>
  )
};

export default Search;



