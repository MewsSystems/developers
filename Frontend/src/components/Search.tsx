import { Input } from 'components/Input'
import React, { useState, useCallback, useRef } from 'react'
import { useTranslation } from 'react-i18next'
import { useTrackEvent } from 'hooks/useTrackEvent'
import { EVENT_CATEGORY, EVENT_ACTION } from 'constants/tracking'

export interface SearchProps {
  value?: string
  onChange?: (value: string) => void
  onClear?: () => void
  onPressEnter?: (value: string) => void
  className?: string
}

export const Search: React.FC<SearchProps> = ({
  value = '',
  onChange,
  onClear,
  onPressEnter,
  className,
}) => {
  const { t } = useTranslation('search')
  const { setTrackEvent } = useTrackEvent()
  const [searchValue, setSearchValue] = useState(value)
  const inputRef = useRef<HTMLInputElement>()

  const handleChange = useCallback(
    ({ target: { value } }: React.ChangeEvent<HTMLInputElement>) => {
      setSearchValue(value)
      onChange && onChange(value)
    },
    [onChange]
  )

  const handleKeyDown = useCallback(
    ({ key }: React.KeyboardEvent<HTMLInputElement>) => {
      if (key === 'Enter' && onPressEnter) {
        onPressEnter(searchValue)

        setTrackEvent({
          eventCategory: EVENT_CATEGORY.SEARCH,
          eventAction: EVENT_ACTION.SEARCH.SUBMIT,
          eventLabel: searchValue,
        })
      }
    },
    [onPressEnter, searchValue, setTrackEvent]
  )

  const handleClear = useCallback(() => {
    setSearchValue('')

    setTrackEvent({
      eventCategory: EVENT_CATEGORY.SEARCH,
      eventAction: EVENT_ACTION.SEARCH.CLEAR,
    })

    onClear && onClear()

    inputRef.current && inputRef.current.focus()
  }, [onClear, setTrackEvent])

  return (
    <Input
      ref={inputRef}
      fullWidth
      allowClear
      inputSize="large"
      type="text"
      placeholder={t('placeholder')}
      value={searchValue}
      onClear={handleClear}
      onChange={handleChange}
      onKeyDown={handleKeyDown}
      className={className}
    />
  )
}
