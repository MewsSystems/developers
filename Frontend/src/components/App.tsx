import React from 'react';
import { BrowserRouter as Router, Route, Switch } from 'react-router-dom';
import Search from './Search/Search';
import Movie from './Movie/Movie';
import { Provider } from 'react-redux';
import createStore from '../store';
import './App.scss';

const store = createStore();

const App = () => {
  return (
    <Provider store={store}>
      <Router>
        <Switch>
          <Route path="/" exact component={Search} />
          <Route path="/movie/:id" render={({ match }) => <Movie id={match.params.id} />} />
        </Switch>
      </Router>
    </Provider>
  );
};

export default App;
