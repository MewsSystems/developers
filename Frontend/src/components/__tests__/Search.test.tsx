import { Search } from 'components/Search'
import { shallow, mount } from 'enzyme'
import React from 'react'

describe('Test Input component', () => {
  it('Tests snapshot', () => {
    expect(shallow(<Search />)).toMatchSnapshot()
  })

  it('Tests functionality', () => {
    const mockOnChange = jest.fn()
    const mockOnEnter = jest.fn()

    const wrapper = mount(
      <Search onChange={mockOnChange} onPressEnter={mockOnEnter} />
    )
    const input = wrapper.find('input')

    input.simulate('change', { event: { target: 'test' } })
    expect(mockOnChange).toBeCalled()

    input.simulate('keydown', { key: 'Enter' })
    expect(mockOnEnter).toBeCalled()
  })
})
