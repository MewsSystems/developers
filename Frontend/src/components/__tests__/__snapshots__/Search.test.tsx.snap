// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Input component Tests snapshot 1`] = `
<ForwardRef
  allowClear={true}
  fullWidth={true}
  inputSize="large"
  onChange={[Function]}
  onClear={[Function]}
  onKeyDown={[Function]}
  placeholder="placeholder"
  type="text"
  value=""
/>
`;
